Refactor empty_line_between_defs cop

unmanbearpig opened the pull request:
Simplified conditionals in empty_line_between_defs cop and extracted a
few methods

That's a follow-up to https://github.com/bbatsov/rubocop/pull/2136

unmanbearpig commented on lib/rubocop/cop/style/empty_line_between_defs.rb:
```diff
@@ -9,27 +9,39 @@ class EmptyLineBetweenDefs < Cop
+        end
+
+        def blank_lines_between?(first_def_node, second_def_node)
+          blank_re = /\A\s*\z/
```
Can someone comment if this is the best way to check if line is blank in Rubocop? 
I think `String#empty?` wouldn't work since it treats lines with whitespace as non-empty. But should we actually allow only-whitespace lines here?

bbatsov replied:
There's `String#blank?` from `powerpack` (a dependency of RuboCop).

unmanbearpig replied:
Thanks! I didn't realize that.

unmanbearpig commented on lib/rubocop/cop/style/empty_line_between_defs.rb:
```diff
@@ -9,27 +9,39 @@ class EmptyLineBetweenDefs < Cop
+        end
+
+        def lines_between_defs(first_def_node, second_def_node)
+          line_range = def_end(first_def_node)..(def_start(second_def_node) - 2)
```
I'm not sure if this is the best way to get lines between `def` nodes, I especially don't like that `- 2` at the end. Can anyone suggest a better way to implement this?

bbatsov replied:
Nothing better comes to mind.

unmanbearpig commented on spec/rubocop/cop/style/empty_line_between_defs_spec.rb:
```diff
@@ -141,6 +141,18 @@
                              '  def b; end'].join("\n"))
   end
 
+  it 'treats lines with whitespaces as blank' do
```
Not sure if we need to support this case.

bbatsov replied:
I'm fine either way. We have a separate cop for trailing whitespace anyways.

unmanbearpig replied:
Got it, then I think it would be better to support it just in case someone disables trailing whitespace cop.

bbatsov commented on spec/rubocop/cop/style/empty_line_between_defs_spec.rb:
```diff
@@ -141,6 +141,18 @@
+              '    end',
+              '  end']
+    inspect_source(cop, source)
+    expect(cop.offenses.size).to eq(0)
```
to be_empty

bbatsov commented on CHANGELOG.md:
```diff
@@ -7,6 +7,7 @@
 * [#1351](https://github.com/bbatsov/rubocop/issues/1351): Allow class emitter methods in `Style/MethodName`. ([@jonas054][])
 * [#2126](https://github.com/bbatsov/rubocop/pull/2126): `Style/RescueModifier` can now auto-correct. ([@rrosenblum][])
 * [#2109](https://github.com/bbatsov/rubocop/issues/2109): Allow alignment with a token on the nearest line with same indentation in `Style/ExtraSpacing`. ([@jonas054][])
+* [#2158](https://github.com/bbatsov/rubocop/pull/2158): Style/EmptyLineBetweenDefs refactoring. ([@unmanbearpig][])
```
Refactorings don't need changelog entries.

unmanbearpig commented:
I've removed the unneeded changelog entry, switched to using `String#blank?` from Powerpack, and changed the spec expectation to call `be_empty` instead of expecting the size to be 0.

bbatsov commented:
:+1: Looks good. Thanks!
